design/design-doc-composition.md

```diff
@@ -0,0 +1,1400 @@
+  writeConnectionSecretToRef:
+    name: sql
+  classSelector:
+    matchLabels:
+      example: "true"
+---
+apiVersion: database.gcp.crossplane.io/v1beta1
+kind: CloudSQLInstanceClass
+metadata:
+  name: example-resource-class
+  labels:
+    example: "true"
+specTemplate:
+  forProvider:
+    databaseVersion: MYSQL_5_6
+    region: us-west2
+    settings:
+      tier: db-n1-standard-1
+      dataDiskType: PD_SSD
+      dataDiskSizeGb: 10
+      ipConfiguration:
+        ipv4Enabled: true
+  writeConnectionSecretsToNamespace: crossplane-system
+  providerRef:
+    name: example
+```
+
+A managed resource is a _high-fidelity_ representation of its corresponding
+external resource. High-fidelity in this context means two things:
+
+* A managed resource maps to exactly one external resource - one API object.
+* A managed resource is as close to a direct translation of its corresponding
+  external API object as is possible without violating [API conventions].
+
+These properties make managed resources - Crossplane's lowest level
+infrastructure primitive - flexible and self documenting. Managed resources in
+and of themselves hold few opinions about _how_ they should be used, and are
+easily related back to the APIs they represent. This provides a solid foundation
+upon which infrastructure operators may build a platform for their application
+operators.
+
+Application operators are typically prevented by [RBAC] from creating and
+modifying managed resources directly; they are instead expected to dynamically
+provision the managed resources they require by submitting a resource claim.
+Crossplane provides claim kinds for common, widely supported resource variants
+like `MySQLInstance` and `KubernetesCluster`. There is a one-to-one relationship
+between claims and the managed resources they bind to; a `KubernetesCluster`
+claim binds to exactly one `GKECluster` managed resource. However, a solitary
+resource is often not particularly useful without supporting infrastructure, for
+example:
+
+* An RDS instance may be inaccessible without a security group.
+* An Azure SQL instance may be inaccessible without a virtual network rule.
+* A GKE, EKS, or AKS cluster (control plane) may not be able to run pods without
+  a node group.
+
+Crossplane infrastructure providers frequently model this supporting
+infrastructure (there is a `SecurityGroup` managed resource, for example), but
+it cannot be dynamically provisioned or bound to a resource claim. Instead a
+cluster operator must statically provision any supporting managed resources
+ahead of time, then author resource classes that reference them. This can be
+limiting:
+
+* Often supporting resources must reference the managed resource they support,
+  for example an Azure `MySQLServerVirtualNetworkRule` must reference the
+  `MySQLServer` it applies to. Dynamically provisioned managed resources such as
+  a `MySQLServer` have non-deterministic names, making it impossible to create a
```

Comment:
I'd like to keep this - I think it is related to composition. Firstly because as you pointed out, reference selection is arguably part of the composition design. It's certainly at least a side effect of the design. Secondly because I think it helps illustrate the shortcomings of the one-to-one claim-to-class relationship that the composition design improves on. Even with selectors there's no way for an application operator to provision the firewall rules they may need.